llvm_passes/HipPromoteInts.cpp

```diff
@@ -0,0 +1,298 @@
+        Value* SrcOp = TruncI->getOperand(0);
+        Value* PromotedSrc = PromotedValues.count(SrcOp) ? PromotedValues[SrcOp] : SrcOp;
+        
+        // Verify the source is actually of our promoted type
+        if (PromotedSrc->getType() != PromotedTy) {
+            PromotedSrc = Builder.CreateZExt(PromotedSrc, PromotedTy);
+        }
+        
+        // Create a new trunc for external users
+        Value* NewTrunc = Builder.CreateTrunc(PromotedSrc, TruncI->getType());
+        errs() << Indent << "  " << *I << "    ============> " << *NewTrunc << "\n";
+        
+        // Store both the promoted and truncated versions
+        PromotedValues[I] = PromotedSrc;  // Use promoted version in our chain
+        Replacements.push_back(Replacement(I, NewTrunc));  // Replace old instruction with new trunc for external users
+    }
+    else if (isa<BinaryOperator>(I)) {
+        BinaryOperator* BinOp = cast<BinaryOperator>(I);
+        bool NeedsPromotion = (BinOp->getType() == NonStdType);
+        
+        Value* LHS = getPromotedValue(BinOp->getOperand(0));
+        Value* RHS = getPromotedValue(BinOp->getOperand(1));
+        
+        Value* NewInst;
+        if (NeedsPromotion) {
+            // Create operation in promoted type
+            NewInst = Builder.CreateBinOp(BinOp->getOpcode(), LHS, RHS);
+        } else {
+            // For operations that should stay in original type
+            if (LHS->getType() != BinOp->getType())
+                LHS = Builder.CreateTrunc(LHS, BinOp->getType());
+            if (RHS->getType() != BinOp->getType())
+                RHS = Builder.CreateTrunc(RHS, BinOp->getType());
+            NewInst = Builder.CreateBinOp(BinOp->getOpcode(), LHS, RHS);
+        }
+        
+        errs() << Indent << "  " << *I << "    ============> " << *NewInst << "\n";
+        PromotedValues[I] = NewInst;
+        Replacements.push_back(Replacement(I, NewInst));
+    }
+    else if (isa<CallInst>(I)) {
+        CallInst* Call = cast<CallInst>(I);
+        // Create a new call with the same operands, but use promoted values where available
+        SmallVector<Value*, 8> NewArgs;
+        for (unsigned i = 0; i < Call->arg_size(); ++i) {
+            Value* Arg = Call->getArgOperand(i);
+            Value* NewArg = PromotedValues.count(Arg) ? PromotedValues[Arg] : Arg;
+            
+            // If the argument needs to match the original type, truncate it
+            if (NewArg->getType() != Call->getArgOperand(i)->getType())
+                NewArg = Builder.CreateTrunc(NewArg, Call->getArgOperand(i)->getType());
+                
+            NewArgs.push_back(NewArg);
+        }
+        
+        CallInst* NewCall = CallInst::Create(Call->getFunctionType(),
+                                           Call->getCalledOperand(),
+                                           NewArgs,
+                                           Call->getName(),
+                                           Call);
```

Comment:
This will create broken code as the called function is not promoted and still have non-standard integer parameters.

Reply:
Can you provide an example IR where this would happen?

Reply:
If a call happens to be promoted, you would end up creating call with promoted argument list to a non-promoted function. Something like:
```
target datalayout = "e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024"
target triple = "spirv64"

define spir_func void @bar_new(i32 noundef %x) local_unnamed_addr #0 {
entry:
  tail call spir_func void @foo(i32 noundef %x) #2
  ret void
}

declare spir_func void @foo(i42 noundef) local_unnamed_addr #1
```
Interestingly, LLVM's verifier does not complain about this, but this will fail in llvm-spirv.

Reply:
This is also a part of the replacement chain. Can you provide an example where we are left with non-std types after the replacement is complete? Or suggest a way to perform the replacemet without intermediary values?